File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    int[][] dp;

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) sum += num;
        if (sum % 2 != 0) return false;

        int target = sum / 2;
        dp = new int[nums.length][target + 1];
        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        return fun(0, 0, nums, target);
    }

    public boolean fun(int index, int currentSum, int[] nums, int target) {
        if (currentSum == target) return true;
        if (index == nums.length || currentSum > target) return false;

        if (dp[index][currentSum] != -1) {
            return dp[index][currentSum] == 1;
        }

        boolean pick = fun(index + 1, currentSum + nums[index], nums, target);
        boolean skip = fun(index + 1, currentSum, nums, target);

        dp[index][currentSum] = (pick || skip) ? 1 : 0;
        return dp[index][currentSum] == 1;
    }
}

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)
        if s % 2 != 0:
            return False
        dp = [[False] * ((s//2)+1) for _ in range(len(nums)+1)]
        dp[0][0] = True
        for i in range(1,len(nums)+1):
            for r in range(s//2+1):
                dp[i][r] = dp[i-1][r]
                if not dp[i][r] and nums[i-1] <= r:
                    dp[i][r] = dp[i][r] or dp[i-1][r - nums[i-1]]
        return dp[len(nums)][s//2]